### PR TITLE
pipewire: 0.3.21 -> 0.3.23

### PR DESCRIPTION
**Pull request by jansol**
New version, new config shenanigans. This release also introduces Pro Audio profiles for applicable soundcards which is nice and all, but also selected by default in this release, which for desktop use is generally not desired. The defaults have been reverted upstream already so it may be reasonable to wait until 0.3.23 before merging this PR.

In the meanwhile I'm currently using pipewire with this PR applied. The config should maybe be refactored into submodules or something. No idea. It works (on my machine) and you can override specific settings from configuration.nix without nuking the rest of the defaults on the side.

Sample configuration.nix, note the quotes in the bluez5 args, pipewire frequently uses periods in config keys and leaving them unquoted makes nix think they are different hierarchy levels.
```nix
  security.rtkit.enable = true;
  services.pipewire = {
    enable = true;
    alsa.enable = true;
    alsa.support32Bit = true;
    pulse.enable = true;
    jack.enable = true;

    media-session.config.bluez-monitor.properties = {
      # MSBC is not expected to work on all headset + adapter combinations.
      "bluez5.msbc-support" = true;
      "bluez5.sbc-xq-support" = true;
    };
  };
```

###### Motivation for this change

https://gitlab.freedesktop.org/pipewire/pipewire/-/releases/0.3.22
https://gitlab.freedesktop.org/pipewire/pipewire/-/releases/0.3.23

###### Things done

**Config file generation is completely redone, comments on that are very much appreciated.**

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jtojnar** [comment]
It would be nice to document how to create the JSON files (e.g. in a `README.md` file in the directory) and also add a test that checks they are up to date.

**Comment by collares** [comment]
Thanks a lot for the update, it's much appreciated! With regards to your comment about waiting for 0.3.23, Fedora's pipewire-0.3.22 packages contains [17 (!) backported patches](https://src.fedoraproject.org/rpms/pipewire/blob/4429acfe32f7bf886396d0863a2f11f340684664/f/pipewire.spec#_54) to fix user-facing problems. If you don't want to wait for 0.3.23, it's probably a good idea to use the same patches.

**Comment by jansol** [comment]
@jtojnar I have no idea how to implement such a test (or any tests) here. Can you point me at the right part of the manual(s) offhand? I guess the best way would be to actually generate the json files as an extra build step and place them in a nix-support output. Then the test could grab any files that are there and diff them against the ones in the module.

@collares that's A LOT of patches for something with bi-weekly releases. At that point I'd rather just grab the latest git revision...

**Comment by jtojnar** [comment]
@jansol Building and diffing is all you need. We already have some consistency checks:

https://github.com/NixOS/nixpkgs/blob/3a2195b7501762b6f218aeb0b9090b4a22d82306/pkgs/development/libraries/pipewire/default.nix#L134-L143

https://github.com/NixOS/nixpkgs/blob/3a2195b7501762b6f218aeb0b9090b4a22d82306/pkgs/development/libraries/pipewire/test-paths.nix#L3-L23

File: nixos/modules/services/desktops/pipewire/pipewire.nix [hunk 1]
```diff
@@ -18,11 +18,26 @@ let
     ln -s "${cfg.package.jack}/lib" "$out/lib/pipewire"
   '';
 
+  prioritizeNativeProtocol = { "context.modules" = { "libpipewire-module-protocol-native" = { _priority = -100; _content = null; }; }; };
+  sessionManagerInvocation = { "context.exec" = { "${cfg.sessionManagerExecutable}" = { args = "${lib.concatStringsSep " " cfg.sessionManagerArguments}"; }; }; };
```
Comment (SuperSandro2000) [inline nixos/modules/services/desktops/pipewire/pipewire.nix:22]:
```suggestion
  prioritizeNativeProtocol = {
    "context.modules" = {
      "libpipewire-module-protocol-native" = {
        _priority = -100; _content = null;
      };
    };
  };

  sessionManagerInvocation = {
    "context.exec" = {
      "${cfg.sessionManagerExecutable}" = {
        args = "${lib.concatStringsSep " " cfg.sessionManagerArguments}";
      };
    };
  };
```

**Comment by jansol** [comment]
No idea if this is even remotely correct.

@collares added the patches while I was at it, although they'll likely be obsolete before I figure those tests out :sweat_smile:

**Comment by collares** [comment]
@jansol That's great, many thanks! I used your commit, some emacs macros and a bit of manual labor to turn this into a bunch of `fetchpatch` calls (see below). Feel free to use it if you want :)

<details>
<summary>fetchpatch version</summary>
```nix
      # Various quality of life improvements that didn't make it into 0.3.22
      (fetchpatch {
        name = "0001-bluez5-include-a2dp-codec-profiles-in-route-profiles.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/df1dbee687c819095a6fddce3b24943f9ac47dbc.patch";
        sha256 = "sha256-MesteaIcfKMr53TdObEfqRyKEgalB1GEWpsTexawPgg=";
      })
      (fetchpatch {
        name = "0001-pulse-server-don-t-use-the-pending_sample-after-free.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/72acd752f68a35c40641262be1c69014170d0734.patch";
        sha256 = "sha256-F3+aVfOYH4SLTwWhdIKuNUo85NBoS+67vkUrcsaldSs=";
      })
      (fetchpatch {
        name = "0005-fix-some-warnings.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/63a3811aa76802e9585e04cf12bf8de617442105.patch";
        sha256 = "sha256-TraIVVU4lE/qfliCTSCdHwoXSq2I4btbNRKjGZjnsuI=";
      })
      (fetchpatch {
        name = "0006-spa-escape-double-quotes.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/6a219092818a7a7e357441256d8070ae77ac2d3a.patch";
        sha256 = "sha256-NUjxOSthe3Jw2LUrO11QnTWTx0PQNBJkQSv3ux58jZw=";
      })
      (fetchpatch {
        name = "0009-bluez5-volumes-need-to-be-distributed-to-all-channel.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/8c5ca000ef7fddd9be9cd2211d880736e97b956d.patch";
        sha256 = "sha256-9NhbYSXlNOYmoPg8tj561QHisaD216fBG7ccNIfoZ64=";
      })
      (fetchpatch {
        name = "0010-bluez5-set-the-right-volumes-on-the-node.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/930b411075a5c1b0d9ddcfe7dd2dbfdc27baab90.patch";
        sha256 = "sha256-xsL92H8wSrBzoxLtraCVxPZdXWTwj+j+O2XlfhcJ7IQ=";
      })
      (fetchpatch {
        name = "0011-bluez5-backend-native-Check-volume-values.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/7a5a94470453e0cbd3a087b1be2aa8960ff43c07.patch";
        sha256 = "sha256-HKWcfkwMmmlEB7l+SETJDsUZ2eLggcodtbNOpqY/wWY=";
      })
      (fetchpatch {
        name = "0012-media-session-don-t-switch-to-pro-audio-by-default.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/84fc63e60168e2c2f43875423899b4ae3bd89a7b.patch";
        sha256 = "sha256-tZwgBMZS+HJd0lrRji3aevbP/yJfvWYK60KMpfiFrOY=";
      })
      (fetchpatch {
        name = "0013-audioconvert-keep-better-track-of-param-changes.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/44919c83fc51aa5e1a22c903e841897c1b4d274c.patch";
        sha256 = "sha256-DIZUfEjUCaK4yY3k7/KLHv1igTKz6Ek4QbWECs1sr9E=";
      })
      (fetchpatch {
        name = "0018-pulse-server-print-encoding-name-in-format_info.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/4b591df145afa316e8e047764f9b98fc10cbacd9.patch";
        sha256 = "sha256-Dr6w6tI4R+6wrl0rVk9Eifw3YleuisVmXzsv+H2jtXc=";
      })
      (fetchpatch {
        name = "0019-pulse-server-handle-unsupported-formats.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/fcf00b3d352d931446a384a768b999520662b599.patch";
        sha256 = "sha256-kQ3vRoojLVTOJi8V4eODkZC8agsla2PjKf6sJnm/1C4=";
      })
      (fetchpatch {
        name = "0021-jack-handle-client-init-error-with-EIO.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/715a3a642a4f1c1f624e0a9dc0394b1819b69f93.patch";
        sha256 = "sha256-DglHcH9pHSWuVxZuLt8J+gDmxOaQSoiThVE2wePfH1I=";
      })
      (fetchpatch {
        name = "0022-pw-cli-always-output-to-stdout.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/91875c1fd8ece6b1b94130c50304b715654f8681.patch";
        sha256 = "sha256-3Sf/Tr39bdBihq/CFborm7GcBWtSvqO3KtRtFBcHcmc=";
      })
      (fetchpatch {
        name = "0024-policy-node-don-t-crash-without-metadata.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/3673265ae20d7b59e89cad6c5238c232796731b2.patch";
        sha256 = "sha256-YRdh1gqBYqa+sbSfHuJZ+Xc6w4QTOOpt/fppFNxsj+8=";
      })
      (fetchpatch {
        name = "0025-bluez5-route-shouldn-t-list-a2dp-profiles-when-not-c.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/a5dc2493dfccfc168cef01162f38f90e986b8af4.patch";
        sha256 = "sha256-zu7zqHIMOcxLAhvelKNWDB+804x+ocCo5FAMUSSOnVo=";
      })
      (fetchpatch {
        name = "0027-jack-apply-PIPEWIRE_PROPS-after-reading-config.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/149319819aa5d4e39770c35a0c00f6c2963e2bf3.patch";
        sha256 = "sha256-SwKQJkg8NIL9D0GAL6hnEgi8kvuToo1bpDm52adtKxA=";
      })
      (fetchpatch {
        name = "0038-jack-add-config-option-to-shorten-and-filter-names.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/d54da879bf5f9fa9009e1ab27e8d52eae3141764.patch";
        sha256 = "sha256-iYOmZTWNODHYWpYGfRTQhBLV1HeFOUvHN3Ta0f1ZS8M=";
      })
      (fetchpatch {
        name = "0046-jack-fix-names-of-our-ports.patch";
        url = "https://gitlab.freedesktop.org/pipewire/pipewire/commit/e340a44a357e4550e169bc8825c6aa8b4ceb76db.patch";
        sha256 = "sha256-rfv2IzgYyNB1s51eRFhJzZCyso6PIEj2qg9fpPpOZ5Q=";
      })
```
</details>

**Comment by collares** [comment]
@jansol My above comment was slightly wrong but I've now updated it. Please make sure you get the most recent version :)

File: pkgs/development/libraries/pipewire/default.nix [hunk 2]
```diff
@@ -72,6 +73,26 @@ let
       ./pipewire-pulse-path.patch
       # Add flag to specify configuration directory (different from the installation directory).
       ./pipewire-config-dir.patch
+
+      # Various quality of life improvements that didn't make it into 0.3.22
```
Comment (Mic92) [inline pkgs/development/libraries/pipewire/default.nix:77]: With this many patches should we just go to a revision on master? Releases are so frequent that we probably get 0.3.23 soon.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 114482` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages marked as broken and skipped:
</summary>
<ul>
<li>libsForQt512.xdg-desktop-portal-kde</li>
<li>libsForQt514.xdg-desktop-portal-kde</li>
</ul>
</details>
<details>
<summary>
1 package blacklisted:
</summary>
<ul>
<li>tests.nixos-functions.nixos-test</li>
</ul>
</details>
<details>
<summary>
26 packages built:
</summary>
<ul>
<li>adapta-gtk-theme</li>
<li>adoptopenjdk-icedtea-web</li>
<li>chrome-gnome-shell</li>
<li>firefox-esr-78-unwrapped (firefoxPackages.firefox-esr-78)</li>
<li>xulrunner (firefox-unwrapped)</li>
<li>gnome3.gnome-control-center (gnome3.gnome_control_center)</li>
<li>gnome3.gnome-remote-desktop</li>
<li>gnome3.gnome-session (gnome3.gnome_session)</li>
<li>gnome3.gnome-shell (gnome3.gnome_shell)</li>
<li>gnome3.gnome-terminal (gnome3.gnome_terminal)</li>
<li>gnome3.gnome-tweaks (gnome3.gnome-tweak-tool)</li>
<li>gnome3.gpaste</li>
<li>gnome3.mutter</li>
<li>gnome3.pomodoro</li>
<li>gnomeExtensions.easyScreenCast</li>
<li>gnomeExtensions.gsconnect</li>
<li>gnomeExtensions.night-theme-switcher</li>
<li>gnomeExtensions.system-monitor</li>
<li>libsForQt5.xdg-desktop-portal-kde (libsForQt515.xdg-desktop-portal-kde ,plasma5Packages.xdg-desktop-portal-kde)</li>
<li>pantheon.elementary-session-settings</li>
<li>pipewire</li>
<li>pulseeffects-pw</li>
<li>slimerjs</li>
<li>xdg-desktop-portal</li>
<li>xdg-desktop-portal-gtk</li>
<li>xdg-desktop-portal-wlr</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
adapta-gtk-theme:
<details>


warning: unclear-gpl
`gpl2` is a deprecated license, check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/data/themes/adapta/default.nix:45:5:

```
   |
45 |     license = with licenses; [ gpl2 cc-by-sa-30 ];
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
</details>
</details>

File: pkgs/development/libraries/pipewire/default.nix (hunk 2 above)
Comment (jansol) [reply pkgs/development/libraries/pipewire/default.nix:77]: That's what I was thinking. I still don't really understand how the module tests should work so chances are 0.3.23 is out before that anyway. Having the patches in here in the meanwhile helps those who want to test the PR in the meanwhile though, so no harm in keeping them around until upstream gets a new version.

**Comment by etircopyh** [comment]
[0.3.23](https://gitlab.freedesktop.org/pipewire/pipewire/-/commit/68f6c75caed047af32320ab4de0c06457457be54) is out.

File: pkgs/development/libraries/pipewire/default.nix (hunk 2 above)
Comment (Mic92) [reply pkgs/development/libraries/pipewire/default.nix:77]: New version available: https://github.com/PipeWire/pipewire/commit/68f6c75caed047af32320ab4de0c06457457be54

Comment (collares) [reply pkgs/development/libraries/pipewire/default.nix:77]: Note that upstream now starts pipewire-media-session via a systemd service, but Fedora opted to just uncomment the exec line in pipewire.conf instead: https://src.fedoraproject.org/rpms/pipewire/blob/rawhide/f/0001-conf-start-media-session-through-pipewire.patch

Comment (jansol) [reply pkgs/development/libraries/pipewire/default.nix:77]: The separate systemd service is much cleaner for the nixos module config and wireplumber seems to have one too, so I'll just use that.

Comment (etircopyh) [reply pkgs/development/libraries/pipewire/default.nix:77]: Ideal option would be for user to be able to choose which session manager he wants to use. Of course pipewire-media-session would be default. But for this to work we should wait for a new WirePlumber release, as 0.3.0 is more than half an year old.

Comment (jansol) [reply pkgs/development/libraries/pipewire/default.nix:77]: Yes that would be optimal. There are some tiny obstacles like no other session managers being packaged at all. But it is something I am trying to prepare for.

**Comment by jansol** [comment]
Got the json files generated so the test should be working now as well. I still don't know how to compare them to the ones in the module though.

**Comment by Mic92** [comment]
Thanks for the effort.

**Comment by kamushadenes** [comment]
Hello,

After this update, I'm getting

```
Mar 09 19:17:56 enterprise pipewire-media-session[20252]: core 0x5621a4246860: can't find protocol 'PipeWire:Protocol:Native': Operation not supported
Mar 09 19:17:56 enterprise pipewire-media-session[20252]: can't start monitor: Operation not supported
```

Still investigating

**Comment by kamushadenes** [comment]
Uninstalling and reinstalling changed the error to now be on `pipewire-pulse`

```
Mar 10 03:04:18 enterprise pipewire-pulse[5160]: context 0x55ff52e25a30: can't load config pipewire-pulse.conf: No such file or directory. Using client.conf fallback
Mar 10 03:04:18 enterprise pipewire-pulse[5160]: context 0x55ff52e25a30: can't load client.conf config: No such file or directory
Mar 10 03:04:18 enterprise pipewire-pulse[5160]: failed to create context: No such file or directory
```

A reboot after that solved the issue.a

**Pull request merged**